datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -1138,27 +1138,44 @@ fn numeric_string_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<D
     }
 }
 
+fn coerce_list_children(lhs_field: &FieldRef, rhs_field: &FieldRef) -> Option<FieldRef> {
+    Some(Arc::new(
+        Arc::unwrap_or_clone(Arc::clone(lhs_field)).with_data_type(comparison_coercion(
+            lhs_field.data_type(),
+            rhs_field.data_type(),
+        )?),
+    ))
+}
+
 /// Coercion rules for list types.
 fn list_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
     use arrow::datatypes::DataType::*;
     match (lhs_type, rhs_type) {
-        (List(_), List(_)) => Some(lhs_type.clone()),
-        (LargeList(_), List(_)) => Some(lhs_type.clone()),
```

Comment:
You can add some test cases of what used to fail and now it doesn't.
For example query unioning list(uint8) with list(int32) where right values don't fit int8 (too big or negative).